soc/arm/st_stm32/stm32wl/power.c

```diff
@@ -76,20 +86,34 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 /* Handle SOC specific activity after Low Power Mode Exit */
 __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 {
-	ARG_UNUSED(substate_id);
-
 	switch (state) {
 	case PM_STATE_SUSPEND_TO_IDLE:
 		LL_LPM_DisableSleepOnExit();
 		LL_LPM_EnableSleep();
+		switch (substate_id) {
+		case 1: /* STOP0 */
+			break;
+		case 2: /* STOP1 */
+			break;
+		case 3: /* STOP2 */
+			/* Reinitialize peripheral functionality */
+			DT_FOREACH_STATUS_OKAY(st_stm32_adc,    STM32_ADC_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_i2c_v2, STM32_I2C_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_spi,    STM32_SPI_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_usart,  STM32_USART_REINIT);
```

Comment:
This is IMO the most controversial part of this work:
- This is performed before `stm32_clock_control_init()` and potentially this could have some impacts in some clock configuration for some peripherals
- This does not take into account actions performed in device's `_pm_action()` function.
- This is not ideal to maintain
- Ideally, there should be a way, in device's `_pm_action()` functions to take into account the PM state we're coming back and act accordingly. Maybe @ceolin can comment on this assertion.

Reply:
Good points, let me try to devise a better approach.

Reply:
After some deeper review of the Device PM documentation and code, I've realized I had some false presumptions about how it worked in relation to System PM. I'd initially thought they worked somewhat independently, but now I see that System PM state changes will trigger Device PM actions.

I think the following would be the cleanest implementation of the reinit logic with Device PM:
1. In the `pm.c` `pm_suspend_devices` and `pm_resume_devices` functions, a `pm_state_info` struct pointer (from `z_cpus_pm_forced_state[CURRENT_CPU]`) is passed to the individual drivers' `*_pm_action` callback, as a (new) third parameter.
2. The drivers will then have access to the System PM power-state that triggered their `*_pm_action` callback.
From there they can make decisions, such as comparing `pm_state_info` with the `power-state` nodes referenced by the `reinit-power-states` phandles, and call the appropriate reinit function, should `PM_DEVICE_ACTION_RESUME` be the triggering action.

The downside is that this is a major change across the Zephyr codebase. I'll need to update all drivers' implementation of the `*_pm_action` callback, so I'd really like @ceolin's input before adding that to this PR.

**Final note:**
Using the drivers' `_pm_action` callback to restore based on system power state is a much cleaner approach, but it does leave open a corner-case for users enabling `CONFIG_PM` without `CONFIG_PM_DEVICE`. In this case, peripherals won't be restored when the system returns from STOP2, I think we should document this somewhere, so that these users don't end up having to chase this down this PR.

Reply:
I don't necessarily think it is a bad idea passing the system state for the device, but as you mentioned that is a major change across the code base and until now that is the first case where we see this needed. Can't you simply use `pm_state_next_get()` in the driver's `pm_action`  to check which power state triggered it ?

This function returns the state selected by the policy and if you call it from a  device `pm_action` it will give which state the SoC will go.

Reply:
After checking the code and seeing that `pm_state_next_get(cpu)` directly returns the passed CPU's `pm_state_info`, I guess that could work. I'd originally written off that function due to it having the word *next* in it, as I'd assumed it would return the state following the current state, e.g. ACTIVE or RUNTIME_IDLE.

However, that would still make correct peripheral functionality dependent on `PM_DEVICE` and `!PM_DEVICE_RUNTIME_EXCLUSIVE`, which is not the the default configuration when enabling simple PM, with our without device PM. I other words it's still lead to broken functionality for the majority of users who don't explicitly disable `PM_RUNTIME_DEVICE_EXCLUSIVE`, or who only use `CONFIG_PM` without device PM enabled.

Could you have a look at my answer to your other comment [below](https://github.com/zephyrproject-rtos/zephyr/pull/59200#discussion_r1245798774).

Reply:
Sorry, not getting your point here. If you don't enable `PM_DEVICE` you don't have device power management support ... in other words no `pm_device_action_run` . No one will trigger device power management.

And if you have `PM_RUNTIME_DEVICE_EXCLUSIVE` enabled, the kernel will not suspend/resume devices you have to it manually.

Reply:
Now uses reworked `pm_notifier` logic from #60998.